### PR TITLE
fix the ambiguity of simultaneous support for multiple backends and supplement tests

### PR DESCRIPTION
**Pull request by zhudada0120**
**Background**

The previous implementation used a function-local static variable: static at::DeviceType at_fork_device_type = device_type;. This logic would "lock in" the first registered device type and ignore all subsequent registration requests — leading to ambiguity: the system appeared to support multiple device types, but only the first registered one would actually work. Following discussions in Issue #166256, we reached a consensus: this ambiguity should be eliminated to support safe forking behavior in scenarios where "multiple backends are initialized simultaneously within a single process".

https://github.com/pytorch/pytorch/blob/d46d8d6f54b15ded4f2483c7bde31be124281ab8/torch/csrc/utils/device_lazy_init.cpp#L79-L92

**Changes**

In device_lazy_init.cpp:
- Add a globally registered-once atfork callback: at_fork_register_once. Unify the atfork handling logic to avoid the single-device-type limitation and multi-callback registration complexity of the old implementation.
- Introduce and use a per-device-type touch flag: at_fork_registered, which records backends that have been accessed/initialized in the parent process.
- In the child process atfork callback, iterate over all device types with at_fork_registered == true: set their "bad fork" flag and reset their lazy initialization requirement.
- Remove the old per-device once_flag declaration (at_fork_once_flags).

**Tests**

Add coverage for:
Poison fork behavior: after forking, initializing the device in the child process fails with a clear RuntimeError.

We conducted a test to validate the outcome of child-process backend device re-initialization following parent-process multi-device initialization, with results aligning with expectations.
image
image

Fixes #166256 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

### CODE REVIEW

**Comment by zhudada0120** [comment]
Hi, @fffrog PTAL, thanks

File: torch/csrc/utils/device_lazy_init.cpp [hunk 1]
```diff
@@ -78,16 +78,24 @@ void set_device_in_bad_fork(at::DeviceType device_type, bool value) {
...
-      if (is_device_lazy_init_supported(at_fork_device_type)) {
-        set_requires_device_init(at_fork_device_type, true);
+      set_device_in_bad_fork(at_fork_device_type_global, true);
+      if (is_device_lazy_init_supported(at_fork_device_type_global)) {
+        set_requires_device_init(at_fork_device_type_global, true);
       }
     });
   });
+  if (device_type != at_fork_device_type_global) {
+    TORCH_CHECK(
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp]:
```suggestion
    TORCH_CHECK(device_type == at_fork_device_type_global)
```

File: test/test_device_lazy_init.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import multiprocessing
+import unittest
+from torch.testing._internal.common_utils import TestCase, run_tests, IS_WINDOWS
+import torch
+
+
+class TestDeviceLazyInit(TestCase):
```
Comment (fffrog) [inline test/test_device_lazy_init.py]: - We can add these tests to this file (test/test_utils.py) without creating a new file.

- These tests need to run on multiple accelerators, such as CUDA, MPS, and XPU, so we can use a function called `instantiate_device_type_tests` to automatically create device-specific test cases.

File: torch/csrc/utils/device_lazy_init.cpp [hunk 3]
```diff
@@ -78,16 +78,24 @@ void set_device_in_bad_fork(at::DeviceType device_type, bool value) {
...
-      set_device_in_bad_fork(at_fork_device_type, true);
-      if (is_device_lazy_init_supported(at_fork_device_type)) {
-        set_requires_device_init(at_fork_device_type, true);
+      set_device_in_bad_fork(at_fork_device_type_global, true);
+      if (is_device_lazy_init_supported(at_fork_device_type_global)) {
+        set_requires_device_init(at_fork_device_type_global, true);
       }
     });
   });
+  if (device_type != at_fork_device_type_global) {
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp]: Because of the constraint `call_once`, the value `at_fork_device_type_glaobal` will remain unchanged, so there is no race condition and it is thread-safe.

File: torch/csrc/utils/device_lazy_init.cpp (hunk 1 above)
Comment (zhudada0120) [reply torch/csrc/utils/device_lazy_init.cpp]: Fixed

File: test/test_utils.py [hunk 4]
```diff
@@ -1011,5 +1011,31 @@ def test_deprecated(self):
...
+                q.put(e)
+
+        ctx = multiprocessing.get_context("fork")
+        q = ctx.Queue()
+        p = ctx.Process(target=child, args=(q,))
+        p.start()
+        p.join()
+        self.assertTrue(not q.empty())
+        exc = q.get()
+        self.assertIsInstance(exc, RuntimeError)
```
Comment (fffrog) [inline test/test_utils.py:1046]: Hey, it is necessary to match the assert strings, just like:

```Python
self.assertRaisesRegex(
            RuntimeError,
            (
                "Cannot re-initialize xxx in forked subprocess. "
                "To use xxx with multiprocessing, you must use the 'spawn' start method"
            ),
)
```

Comment (zhudada0120) [reply test/test_utils.py:1046]: Thanks for the advice. I have added an assertion for error_msg. PTAL.

File: test/test_utils.py [hunk 5]
```diff
@@ -53,10 +54,7 @@
 # sharding on sandcastle. This line silences flake warnings
 load_tests = load_tests  # noqa: PLW0127
 
-device_type = (
-    acc.type if (acc := torch.accelerator.current_accelerator(True)) else "cpu"
-)
-TEST_GPU = torch.xpu.is_available() or torch.cuda.is_available()
+HAS_CUDA = torch.cuda.is_available()
```
Comment (albanD) [inline test/test_utils.py]: Why this change?

File: test/test_utils.py [hunk 6]
```diff
@@ -53,10 +54,7 @@
 # sharding on sandcastle. This line silences flake warnings
 load_tests = load_tests  # noqa: PLW0127
```
Comment (fffrog) [inline test/test_utils.py:56]: @zhudada0120, please restore those changes, which will needed in other test suits.

File: test/test_utils.py (hunk 5 above)
Comment (zhudada0120) [reply test/test_utils.py]: Thank you for your suggestion. I have reverted this change.Thanks.

**Comment by zhudada0120** [comment]
Hi @albanD , I have fixed the multi-device support bug. PTAL.Thanks.

**Comment by fffrog** [comment]
Hi, @zhudada0120.

The current changes may cause some issues in certain scenarios, so we may need to add support for multiple devices to the `register_fork_handler_for_device_init` function first. As for adhering to the design principle of using only one accelerator per program, we can address that later.

**Comment by albanD** [comment]
btw all CI is disabled on the repo due to the ongoing npm worm for security. See slack announcement channel or the issue here for details.

**Comment by fffrog** [comment]
> The current changes may cause some issues in certain scenarios, so we may need to add support for multiple devices to the register_fork_handler_for_device_init function first. As for adhering to the design principle of using only one accelerator per program, we can address that later.

Hi @zhudada0120, could you help refactor this PR to support multiple accelerators simultaneously? In other words, we just need to fix bugs in scenarios with multiple accelerators.

**Comment by zhudada0120** [comment]
Hi @fffrog @albanD, Thanks for the suggestion. I'll complete the modifications as soon as possible.

File: torch/csrc/utils/device_lazy_init.cpp [hunk 7]
```diff
@@ -16,6 +17,8 @@ std::array<bool, at::COMPILE_TIME_MAX_DEVICE_TYPES> is_initialized{};
 std::array<bool, at::COMPILE_TIME_MAX_DEVICE_TYPES> is_in_bad_fork{};
 std::array<c10::once_flag, at::COMPILE_TIME_MAX_DEVICE_TYPES>
     at_fork_once_flags{};
+std::array<bool, at::COMPILE_TIME_MAX_DEVICE_TYPES> at_fork_registered{};
+c10::once_flag at_fork_register_once{};
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp:18]: Why do we need this one?

File: torch/csrc/utils/device_lazy_init.cpp [hunk 8]
```diff
@@ -77,13 +78,18 @@ void set_device_in_bad_fork(at::DeviceType device_type, bool value) {
...
-        set_requires_device_init(at_fork_device_type, true);
+      for (int i = 0; i < static_cast<int>(at::COMPILE_TIME_MAX_DEVICE_TYPES); ++i) {
+        if (!at_fork_registered[i]) {
+          continue;
+        }
+        auto dt = static_cast<at::DeviceType>(i);
+        set_device_in_bad_fork(dt, true);
+        if (is_device_lazy_init_supported(dt)) {
+          set_requires_device_init(dt, true);
+        }
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp:92]: Hey, thank you for updating this.

It seems like the changes is not very clear and still have other problems.

- We need to support multiply device simultanerous, so we don`t need `at_fork_register_once`
- We need to put `at_fork_registered[static_cast<int>(device_type)] = true; into block of `c10::call_once` to keep consistence in race situation.
- We need to aviod multiple initialzation.

File: torch/csrc/utils/device_lazy_init.cpp [hunk 9]
```diff
@@ -77,13 +78,18 @@ void set_device_in_bad_fork(at::DeviceType device_type, bool value) {
 // Should be called before the first device runtime call.
 void register_fork_handler_for_device_init(at::DeviceType device_type) {
 #ifndef WIN32
-  auto& flag = at_fork_once_flags[static_cast<int>(device_type)];
-  c10::call_once(flag, [device_type]() {
-    static at::DeviceType at_fork_device_type = device_type;
+  at_fork_registered[static_cast<int>(device_type)] = true;
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp:80]: Move this one into c10::call_once

File: torch/csrc/utils/device_lazy_init.cpp [hunk 10]
```diff
@@ -77,13 +78,18 @@ void set_device_in_bad_fork(at::DeviceType device_type, bool value) {
...
+  c10::call_once(at_fork_register_once, []() {
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp:81]: auto& flag = at_fork_once_flags[static_cast<int>(device_type)];
c10::call_once(flag, [device_type]()

File: torch/csrc/utils/device_lazy_init.cpp (hunk 7 above)
Comment (zhudada0120) [reply torch/csrc/utils/device_lazy_init.cpp:18]: at_fork_register_once is a flag used to ensure that the child process callback is registered only once, and at_fork_registered is used to record which devices have been initialized in the current process.

File: torch/csrc/utils/device_lazy_init.cpp [hunk 11]
```diff
@@ -77,13 +78,18 @@ void set_device_in_bad_fork(at::DeviceType device_type, bool value) {
...
     pthread_atfork(nullptr, nullptr, []() {
-      set_device_in_bad_fork(at_fork_device_type, true);
-      if (is_device_lazy_init_supported(at_fork_device_type)) {
-        set_requires_device_init(at_fork_device_type, true);
+      for (int i = 0; i < static_cast<int>(at::COMPILE_TIME_MAX_DEVICE_TYPES); ++i) {
+        if (!at_fork_registered[i]) {
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp:85]: We need to distinguish which devices have been registered before.

File: torch/csrc/utils/device_lazy_init.cpp [hunk 12]
```diff
@@ -77,13 +78,18 @@ void set_device_in_bad_fork(at::DeviceType device_type, bool value) {
...
-  auto& flag = at_fork_once_flags[static_cast<int>(device_type)];
-  c10::call_once(flag, [device_type]() {
-    static at::DeviceType at_fork_device_type = device_type;
+  at_fork_registered[static_cast<int>(device_type)] = true;
+  c10::call_once(at_fork_register_once, []() {
     pthread_atfork(nullptr, nullptr, []() {
-      set_device_in_bad_fork(at_fork_device_type, true);
-      if (is_device_lazy_init_supported(at_fork_device_type)) {
-        set_requires_device_init(at_fork_device_type, true);
+      for (int i = 0; i < static_cast<int>(at::COMPILE_TIME_MAX_DEVICE_TYPES); ++i) {
```
Comment (fffrog) [inline torch/csrc/utils/device_lazy_init.cpp]: Moreover, I don't know how to implement it more clearly or in a more understandable way, although I don`t think using a loop is a good way.

**Comment by fffrog** [comment]
> Hi @fffrog, I have modified the code to enable simultaneous support for multiple accelerators.PTAL.thanks.

@zhudada0120 Hey, thank you very much for updating this.

It look good to me, please wait approval from @albanD

**Review by fffrog — APPROVED** [review]
SGTM, thank you.

**Review by albanD — COMMENTED** [review]
Thanks for updating!
Let me just double check it does pass with some weird internal config.

**Comment by albanD** [comment]
@pytorchbot merge

Ok it all looks good!
Thanks for taking the time to update this!

**Review by guangyey — APPROVED** [review]
Thanks!

File: test/test_utils.py [hunk 13]
```diff
@@ -1020,5 +1021,36 @@ def test_deprecated(self):
...
+        pattern = (
+            r"Cannot re-initialize .* in forked subprocess\. "
+            r"To use .* with multiprocessing, you must use the 'spawn' start method"
+        )
+        self.assertIsInstance(exc, RuntimeError)
+        self.assertRegex(str(exc), pattern)
+
+
+instantiate_device_type_tests(
+    TestDeviceLazyInit, globals(), except_for=["cpu"], allow_xpu=True
```
Comment (guangyey) [inline test/test_utils.py:1051]: Thanks. I add ciflow/xpu to test this.

Comment (fffrog) [reply test/test_utils.py:1051]: Hey @guangyey, this failure doesn't seem to be related to this PR, please help to doubel check

Comment (guangyey) [reply test/test_utils.py:1051]: Sorry for your inconvenience, I rebase this PR to re-trigger XPU CI.

**Review by albanD — COMMENTED** [review]
Yes this is ready to go. Do skip flaky CI as well!

**Comment by guangyey** [comment]
Remove xpu ciflow label and retrigger CI to not block this PR.
@pytorchbot rebase

**Review by albanD — APPROVED** [review]
